File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceElementArrayChainTickers.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceElementArrayChainTickers.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.ArrayList;
import java.util.Calendar;
import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class ReferenceElementArrayChainTickers extends ReferenceElementArray
{
	public ReferenceElementArrayChainTickers(String underlier, int numPoints, String strDtExp, OptionalityEnum optionality) throws Exception
	{
		super("CHAIN_TICKERS", new ArrayList<Element>());
		
		int indexSpace = underlier.indexOf(' ');
		String ticker = underlier.substring(0, indexSpace);
		
		Datetime dtExp;
		if(strDtExp == null)
		{
			dtExp = new Datetime();
			dtExp.calendar().add(Calendar.MONTH, 1);
			dtExp.calendar().set(Calendar.DAY_OF_MONTH, 20); //assume the 20th of the month
		}
		else if(strDtExp.length() == 8)
		{
			dtExp = com.bloomberglp.blpapi.DisplayFormats.HistRef_TryParseInput(strDtExp);
		}
		else if(strDtExp.length() == 6)
		{
			dtExp = com.bloomberglp.blpapi.DisplayFormats.HistRef_TryParseInput(strDtExp + "20");
		}
		else if (strDtExp.endsWith("F"))
		{
			String strNumMonths = strDtExp.substring(0, strDtExp.length() - 2);
			int numMonths = Integer.parseInt(strNumMonths);
			Calendar c = Calendar.getInstance();
			c.set(Calendar.HOUR, 0);
			c.set(Calendar.MINUTE, 0);
			c.set(Calendar.SECOND, 0);
			c.add(Calendar.MONTH, numMonths);
			c.set(Calendar.DAY_OF_MONTH, 20);
			dtExp = new Datetime(c);
		}
		else
		{
			throw new Exception("BEmu.ReferenceDataRequest.ElementReferenceArrayChainTickers: unable to determine the dtExp");
		}

        int strike = RandomDataGenerator.strike();
        for (int count = 0; count < numPoints; count++, strike += 5)
        {
        	Element elm = new ReferenceElementArrayChainTickersItem(ticker, dtExp, optionality, strike);        	
        	super._values.add(elm);
        }
	}
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("CHAIN_TICKERS"));
    }
}

File: BEmu/ReferenceDataRequest/ReferenceElementArrayChainTickers.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="ReferenceDataRequest/ReferenceElementArrayChainTickers.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ReferenceElementArrayChainTickers : ReferenceElementArray
    {
        internal enum OptionalityEnum { put = 'P', call = 'C' }

        public override SchemaTypeDefinition TypeDefinition { get { return new SchemaTypeDefinition(this.Datatype, new Name("CHAIN_TICKERS")); } }

        internal ReferenceElementArrayChainTickers(string underlier, uint numPoints, string strDtExp, OptionalityEnum optionality) : base("CHAIN_TICKERS", new List<Element>())
        {
            //assume that the underlier takes the form "ABC US EQUITY"
            string ticker = underlier.Substring(0, underlier.IndexOf(' '));
            
            DateTime dtExp;
            if(Types.Extensions.IsNullOrWhiteSpace(strDtExp))
                dtExp = DateTime.Today.AddMonths(1).AddDays(-DateTime.Today.Day).AddDays(20); //assume the 20th of the month
            else if (strDtExp.Length == 8)
            {
                if (!Types.DisplayFormats.HistoricalOrReferenceRequests.TryParseInput(strDtExp, out dtExp))
                    throw new NotImplementedException("BEmu.ReferenceDataRequest.ElementReferenceArrayChainTickers: unable to determine the dtExp (10)");
            }
            else if (strDtExp.Length == 6)
            {
                if (!Types.DisplayFormats.HistoricalOrReferenceRequests.TryParseInput(strDtExp + "20", out dtExp)) //assume the 20th of the month
                    throw new NotImplementedException("BEmu.ReferenceDataRequest.ElementReferenceArrayChainTickers: unable to determine the dtExp (20)");
            }
            else if (strDtExp[strDtExp.Length - 1] == 'F')
            {
                uint numMonths = uint.Parse(strDtExp.Substring(0, strDtExp.Length - 1));
                dtExp = DateTime.Today.AddMonths((int)numMonths).AddDays(-DateTime.Today.Day).AddDays(20); //assume the 20th of the month
            }
            else
                throw new NotImplementedException("BEmu.ReferenceDataRequest.ElementReferenceArrayChainTickers: unable to determine the dtExp");

            int strike = Types.RandomDataGenerator.Strike();
            for (int count = 0; count < numPoints; count++, strike += 5)
            {
                //string optionTicker = string.Format("{0} US {1:MM/dd/yy} {2}{3}", ticker, dtExp, (char)optionality, strike);

                Element elm = new ReferenceElementArrayChainTickersItem(ticker, dtExp, optionality, strike);

                base._values.Add(elm);
            }
        }
    }
}
